### PR TITLE
Do not compute JSON Schema default when plain serializers are used with `when_used` set to `'json-unless-none'` and the default value is `None`

### LINKED ISSUE #10118: `PlainSerializer` not respecting `when_used` field during generation of json schema
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

We've a `PlainSerializer` e.g.:

```python
Decimal = Annotated[
    _Decimal, PlainSerializer(lambda x: float(x), return_type=float, when_used='json-unless-none')
]

class MyModel(BaseModel):
  my_decimal: Decimal = Field(default=None)

my_model = MyModel(my_decimal=Decimal(1.0))
logger.info("schema {schema}", schema=my_model.model_json_schema(mode='serialization'))
```
which is not None-safe, however the lambda will be called during the schema generation

The schema generattion fails, cause the serialisation lambda will be called either way in this snippet in pydantic (json_schema.py:1065):

```python
if self.mode == 'serialization':
            if (
                (ser_schema := schema['schema'].get('serialization', {}))
                and (ser_func := ser_schema.get('function'))
                and ser_schema.get('type') == 'function-plain'  # type: ignore
                and ser_schema.get('info_arg') is False  # type: ignore
            ):
                default = ser_func(default)  # type: ignore
```

this started after upgrading these packages:

```
  - Updating pydantic-core (2.18.4 -> 2.20.1)
  - Updating pydantic (2.7.4 -> 2.8.2)
```

### Example Code

```Python
Decimal = Annotated[
    _Decimal, PlainSerializer(lambda x: float(x), return_type=float, when_used='json-unless-none')
]

class MyModel(BaseModel):
  my_decimal: Decimal = Field(default=None)

my_model = MyModel(my_decimal=Decimal(1.0))
logger.info("schema {schema}", schema=my_model.model_json_schema(mode='serialization'))
```
```


### Python, Pydantic & OS Version

```Text
python -c "import pydantic.version; print(pydantic.version.version_info())"                                                    ✔  3.11.0   11:35:45
             pydantic version: 2.8.2
        pydantic-core version: 2.20.1
          pydantic-core build: profile=release pgo=true
                 install path: /Users/.../.venv/lib/python3.11/site-packages/pydantic
               python version: 3.11.0 (main, Nov 18 2022, 12:41:05) [Clang 14.0.0 (clang-1400.0.29.102)]
                     platform: macOS-14.6.1-arm64-arm-64bit
             related packages: mypy-0.931 fastapi-0.101.0 typing_extensions-4.12.2 pydantic-settings-2.4.0
                       commit: unknown
```

### PR DESCRIPTION
**Pull request by Viicos**
Fixes #10118

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

File: tests/test_json_schema.py [hunk 1]
```diff
@@ -6257,6 +6257,22 @@ class Model(BaseModel):
...
+    class Model(BaseModel):
+        custom_decimal: Annotated[
+            Decimal | None, PlainSerializer(lambda x: float(x), when_used='json-unless-none', return_type=float)
+        ] = None
+
+    assert Model.model_json_schema(mode='serialization') == {
+        'properties': {'custom_decimal': {'default': None, 'title': 'Custom Decimal', 'type': 'number'}},
+        'title': 'Model',
+        'type': 'object',
+    }
```
Comment (Viicos) [inline tests/test_json_schema.py:6273]: This revealed another issue with the JSON Schema: `type` should be `'anyOf': [{'type': 'number'}, {'type': 'null'}]}`.

When the JSON Schema for our field is created, `schema` is of type `default`, and `schema['schema']` is of type `nullable`:

https://github.com/pydantic/pydantic/blob/0c97049bd40ac95cce7bbebad8369b9426af4b09/pydantic/json_schema.py#L1022-L1031

However, `self.generate_inner(schema['schema']`) will not go through `nullable_schema` as we would expect. As there's a `serialization` key, `ser_schema` is being used:

https://github.com/pydantic/pydantic/blob/0c97049bd40ac95cce7bbebad8369b9426af4b09/pydantic/json_schema.py#L492-L494

This makes sense, however, the issue is `ser_schema` kind of assumes `when_used` is `always`. I'll see if I can fix this in a follow up PR.

File: pydantic/json_schema.py [hunk 2]
```diff
@@ -1043,17 +1043,17 @@ def default_schema(self, schema: core_schema.WithDefaultSchema) -> JsonSchemaVal
...
-                and ser_schema.get('type') == 'function-plain'  # type: ignore
-                and ser_schema.get('info_arg') is False  # type: ignore
-            ):
-                default = ser_func(default)  # type: ignore
+        if (
+            self.mode == 'serialization'
+            and (ser_schema := schema['schema'].get('serialization'))
+            and (ser_func := ser_schema.get('function'))
+            and ser_schema.get('type') == 'function-plain'
+            and not ser_schema.get('info_arg')
```
Comment (Viicos) [inline pydantic/json_schema.py:1053]: Switching for `and ser_schema.get('info_arg') is False` releaved an issue in the tests, that I will fix in this PR.

File: pydantic/json_schema.py [hunk 3]
```diff
@@ -1043,17 +1043,30 @@ def default_schema(self, schema: core_schema.WithDefaultSchema) -> JsonSchemaVal
...
+                # (assuming `validate_default` is enabled). However, we can't perform
+                # such validation during JSON Schema generation so we don't support
+                # this pattern for now.
+                # (One example is when using `foo: ByteSize = '1MB'`, which validates and
+                # serializes as an int. In this case, `ser_func` is `int` and `int('1MB')` fails).
+                self.emit_warning(
+                    'non-serializable-default',
+                    f'Unable to serialize value {default!r} with the plain serializer; excluding default from JSON schema',
+                )
+                return json_schema
```
Comment (sydney-runkle) [inline pydantic/json_schema.py:1069]: Could we do something like this?

https://github.com/pydantic/pydantic/blob/0c97049bd40ac95cce7bbebad8369b9426af4b09/pydantic/json_schema.py#L2017-L2044

I feel like we shouldn't be skipping that logic here, we could maybe call that function?

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
This patch looks alright to me - I've left one comment that I think might simplify things!

File: pydantic/json_schema.py (hunk 3 above)
Comment (Viicos) [reply pydantic/json_schema.py:1069]: I don't think it makes sense to do so, or we may end up with a wrong default value in the JSON Schema (in the related test fixed, `'1MB'` was set as the default value in the JSON Schema, but this isn't true as the default is `1000000`). In my opinion, it's better to have no value than a wrong one.

However, we emit a warning here, so I'm ok to call `encode_default`.

Comment (sydney-runkle) [reply pydantic/json_schema.py]: That's fair - I wonder if we could use a similar pattern though + attempt to validate the default + include that in the JSON schema. Feels messy to do validation here, I admit.

I'm not opposed to the warning you have in place, I think that's good!

Comment (sydney-runkle) [reply pydantic/json_schema.py]: > That's fair - I wonder if we could use a similar pattern though + attempt to validate the default + include that in the JSON schema. Feels messy to do validation here, I admit.

I think we can do this in a separate PR, if someone asks for it. I'm not keen to go ahead and do this now.

Change looks good to me, thanks for including the helpful warning.

**Review by sydney-runkle — APPROVED** [review]
Looks good @Viicos, nice work

File: pydantic/json_schema.py [hunk 4]
```diff
@@ -1043,17 +1043,30 @@ def default_schema(self, schema: core_schema.WithDefaultSchema) -> JsonSchemaVal
...
+            except Exception:
+                # It might be that the provided default needs to be validated (read: parsed) first
+                # (assuming `validate_default` is enabled). However, we can't perform
+                # such validation during JSON Schema generation so we don't support
+                # this pattern for now.
+                # (One example is when using `foo: ByteSize = '1MB'`, which validates and
+                # serializes as an int. In this case, `ser_func` is `int` and `int('1MB')` fails).
+                self.emit_warning(
+                    'non-serializable-default',
+                    f'Unable to serialize value {default!r} with the plain serializer; excluding default from JSON schema',
```
Comment (sydney-runkle) [inline pydantic/json_schema.py:1067]: One nit, this isn't  always a user defined plain serializer, which might be confusing. We could change the wording of this a bit, but I'm not super worried about it, so will go ahead and merge this.

**Pull request merged**